chia/full_node/full_node.py

```diff
@@ -1162,17 +1198,46 @@ async def add_block_batch(
         self,
         all_blocks: List[FullBlock],
         peer_info: PeerInfo,
-        fork_point: Optional[uint32],
+        fork_info: Optional[ForkInfo],
         wp_summaries: Optional[List[SubEpochSummary]] = None,
     ) -> Tuple[bool, Optional[StateChangeSummary], Optional[Err]]:
         # Precondition: All blocks must be contiguous blocks, index i+1 must be the parent of index i
         # Returns a bool for success, as well as a StateChangeSummary if the peak was advanced
 
         blocks_to_validate: List[FullBlock] = []
         for i, block in enumerate(all_blocks):
-            if not self.blockchain.contains_block(block.header_hash):
+            header_hash = block.header_hash
+            if not await self.blockchain.contains_block_from_db(header_hash):
                 blocks_to_validate = all_blocks[i:]
                 break
+
+            if fork_info is None:
+                continue
+            # the below section updates the fork_info object, if
+            # there is one.
+
+            # TODO: it seems unnecessary to request overlapping block ranges
+            # when syncing
+            if block.height <= fork_info.peak_height:
```

Comment:
this is *ignoring* blocks that are below our current peak. i.e. blocks we've already validated.
The main reason this matters here is since we accrue fork state in `fork_info`. We can't include duplicates `additions` and `removals`.

Removing this check, makes the `test_full_node.py::test_long_reorg_nodes` fail like this:
```
19:21:25 chia.full_node.full_node: ERROR sync from fork point failed: AssertionError:
Traceback (most recent call last):
  File "/Users/arvid/Documents/dev/2/chia-blockchain/chia/util/log_exceptions.py", line 20, in log_exceptions
    yield
  File "/Users/arvid/Documents/dev/2/chia-blockchain/chia/full_node/full_node.py", line 1138, in sync_from_fork_point
    await asyncio.gather(fetch_task, validate_task)
  File "/Users/arvid/Documents/dev/2/chia-blockchain/chia/full_node/full_node.py", line 1105, in validate_block_batches
    success, state_change_summary, err = await self.add_block_batch(
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/arvid/Documents/dev/2/chia-blockchain/chia/full_node/full_node.py", line 1238, in add_block_batch
    await self.blockchain.advance_fork_info(block, fork_info)
  File "/Users/arvid/Documents/dev/2/chia-blockchain/chia/consensus/blockchain.py", line 218, in advance_fork_info
    assert fork_info.peak_height <= block.height - 1
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

Reply:
I didn't dig into this very deeply, but it appeared that when we request batches of blocks, we end up adding the last block twice. To me it seemed like the request might be *inclusive* (and request 33 blocks) instead of 32. (hence my TODO comment above). I'm not so sure this is actually what's happening anymore, but we end up adding the last block in the batch twice.

Reply:
i dont think we want to do things we dont fully understand here, it looks to me like we are just patching things without knowing really why

Reply:
I believe this issue has been resolved by https://github.com/Chia-Network/chia-blockchain/pull/16792 but I didn't rebase again after that landed. This logic was here all along though, we used to support adding blocks that we already had, and this ensures we still do.

Reply:
the key difference is that when we update the `ForkInfo` object based on the new object we're adding, it's important to not add the same block twice. This isn't really a concern in the existing code.